wp-env: Add support for different options in each environment

noahtallen opened the pull request:
## Description
Allows the user to change the options for any WordPress environment in wp-env separately. This refactors the way the config works internally by putting all service-level configuration options into `env.<service>`.

_Note: One subtle difference with parsing behavior is that wp-config values are now merged instead of overwritten._

For example:
```
const parsedConfig = {
  ...uniqueOptions, // like dockerComposeConfigPath, workDirectoryPath, etc)
  env {
    development: WPServiceConfig,
    tests: WPServiceConfig,
  },
}
```

WPServiceConfig is just something like:

```
const defaultEnvConfig = {
	core: null,
	plugins: [],
	themes: [],
	port: 8888,
	config: {
		WP_DEBUG: true,
		SCRIPT_DEBUG: true,
		WP_TESTS_DOMAIN: 'localhost',
	},
	mappings: {},
};
```

the `WPServiceConfig` contains all options applicable to the service. For example, if .wp-env.json looks like:
```json
{
  "plugins": [ "./" ],
  "env": {
    "tests": {
      "themes": [ "./some-theme" ]
    }
  }
}
```

the config will be merged like so:
```js
const config = {
  // ... common config options, like workDir
  env: {
    development: {
      // ... other service-level options like port
      plugins: [ "./" ],
      themes: [],
    },
    tests: {
      // ... other service-level options like port
      plugins: [ "./" ],
      themes: [ "./some-theme" ],
    },
  },
}
```

As you can see, the base-level plugins option is added to both development and tests.

Practically, in `.wp-env.json`, the only thing that changes is that you can override any of the existing options inside of `env.development` or `env.tests`.

## Todo:
- [x] Fix wp-env unit tests
- [x] Update Travis config which was working around this issue

## How has this been tested?
Unit tests, travis, and local testing in my own environment

## Types of changes
new feature/refactor

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

closes #22514

noahtallen committed:
Separate config for different options

noahtallen commented on packages/env/lib/config.js:
```diff
@@ -62,200 +77,281 @@ module.exports = {
+ * @param {string} workDirectoryPath Absolute path to the environment work dir.
+ * @return {Object} the validated and parsed config object.
+ */
+function parseConfig( config, workDirectoryPath ) {
```
`parseConfig` now validates a `WPServiceConfig`. this way it can be reused for each environment override. (need to clean up comments around this)

noahtallen commented on packages/env/lib/config.js:
```diff
@@ -62,200 +77,281 @@ module.exports = {
+ *
+ * @param {Object} rawConfig config right after being read from a file.
+ */
+function withBackCompat( rawConfig ) {
```
with the new format, it doesn't make sense to have both `port` and `testsPort`. So each `WPServiceConfig` now has one `port` property. This function makes sure that existing `.wp-env.json` files continue working.

epiqueras commented:
This is looking great! Let us know when it's ready for a pass :)

aduth commented on packages/env/lib/build-docker-compose-config.js:
```diff
@@ -7,16 +7,16 @@ const path = require( 'path' );
 
 /**
  * @typedef {import('./config').Config} Config
+ * @typedef {import('./config').WPServiceConfig} WPServiceConfig
```
I don't believe it's valid to define multiple `@typedef` in the same JSDoc comment.

At some point it would be good to opt-in the package to type-checking (#18838) to both use and validate these type definitions. It feels like a good candidate package for type-checking.

noahtallen replied:
Sounds great to me. I wish it could be typescript; that would make this a lot easier to work with.

noisysocks replied:
I started doing this in the [`try/wp-env-typescript`](https://github.com/WordPress/gutenberg/compare/try/wp-env-typescript) branch a little while ago if you'd like to finish it. It wasn't difficult, just very tedious because a lot of the external packages we're using have incomplete or incorrect type definitions.

epiqueras replied:
Ahh, that's the only annoying part of using TS.

noahtallen commented:
My goal is to get this over the finish line this week. It's taken more time than expected to support multiple core sources. There are some expectations around how the core source is configured that might not always be true any more, so I'm working on solving that

noahtallen committed:
Happy little bug fixes

noahtallen commented:
@epiqueras this is ready for a look. I think some parts of it could be cleaned up more (e.g. `config.js` should be split into more files), but it should be working now and ready for a ✅ on the general approach

epiqueras commented:
>ready for a ✅ on the general approach

Looks good 👍

noahtallen committed:
Split config into multiple files in a directory

noahtallen committed:
Split out function

noahtallen committed:
Move download sources function

noahtallen commented:
@epiqueras this is ready for a thorough look. since config.js was getting pretty large, I moved it into its own directory and split out the functions into a few more files. Everything seems to be working pretty well. The remaining task is to update the tests & docs, but it'd be nice to get feedback on the code

epiqueras commented on packages/env/lib/config/config.js:
```diff
@@ -0,0 +1,311 @@
+		environments.forEach( ( envKey ) => {
+			try {
+				const baseUrl = new URL(
+					config.env[ envKey ].config[ configKey ]
```
What happens when this is undefined?

epiqueras commented on packages/env/lib/config/config.js:
```diff
@@ -0,0 +1,311 @@
+/**
+ * Gets the `wp-env` home directory in which generated files are created.
+ *
+ * By default, '~/.wp-env/'. On Linux, '~/wp-env/'. Can be overriden with the
```
```suggestion
 * By default, '~/.wp-env/'. On Linux with snap packages, '~/wp-env/'. Can be overriden with the
```

epiqueras commented on packages/env/lib/config/config.js:
```diff
@@ -0,0 +1,311 @@
+}
+
+module.exports = {
+	ValidationError,
```
Why is this reexporting `ValidationError`?

epiqueras commented on packages/env/lib/config/index.js:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Internal dependencies
+ */
+const { ValidationError, readConfig } = require( './config' );
```
`ValidationError` could be imported from the validation file.

epiqueras commented on packages/env/lib/config/read-raw-config-file.js:
```diff
@@ -0,0 +1,62 @@
+		rawConfig.env = {
+			...( rawConfig.env || {} ),
+			tests: {
+				...( rawConfig.env && rawConfig.env.tests
+					? rawConfig.env.tests
+					: {} ),
```
```suggestion
				...( rawConfig?.env?.tests || {} ),
```

epiqueras commented on packages/env/lib/config/validate-config.js:
```diff
@@ -0,0 +1,207 @@
+/**
+ * Parses and validates a config object.
+ *
+ * @param {Object} config A raw config object to parse
```
```suggestion
 * @param {Object} config A parsed config object to validate
```

epiqueras commented on packages/env/lib/config/validate-config.js:
```diff
@@ -0,0 +1,207 @@
+ * @param {Object} options
+ * @param {string} options.workDirectoryPath Path to the work directory located in ~/.wp-env.
+ * @param {string} options.environment Environment name for the service we are parsing.
+ * @return {WPServiceConfig} validated and parsed service-level configuration.
```
```suggestion
 * @return {WPServiceConfig} Validated and parsed service-level configuration.
```

epiqueras commented on packages/env/lib/config/validate-config.js:
```diff
@@ -0,0 +1,207 @@
+	for ( const [ wpDir, localDir ] of Object.entries( config.mappings ) ) {
+		if ( ! localDir || typeof localDir !== 'string' ) {
+			throw new ValidationError(
+				`Invalid .wp-env.json: "mapping.${ wpDir }" should be a string.`
```
```suggestion
				`Invalid .wp-env.json: "mappings.${ wpDir }" should be a string.`
```

epiqueras commented on packages/env/lib/download-sources.js:
```diff
@@ -17,26 +17,73 @@ const rimraf = util.promisify( require( 'rimraf' ) );
+ *
+ * @param {Config} config The wp-env configuration object.
+ * @param {Object} spinner The spinner object to show progress.
+ * @return {Promise[]} An array of promises for the downlad tasks.
```
```suggestion
 * @return {Promise[]} An array of promises for the download task.
```

epiqueras commented on packages/env/lib/wordpress.js:
```diff
@@ -180,6 +179,64 @@ async function resetDatabase(
+	for ( const env in envs ) {
+		// If one does not have a core source but the other does.
+		if (
+			( ! env.coreSource && env.coreSource ) ||
```
```suggestion
			( ! coreSource && env.coreSource ) ||
```

noahtallen committed:
Clarify several jsdoc comments

noahtallen committed:
Logic errors, syntax sugar, and clarified comments

noahtallen committed:
Revert nullish operator

noahtallen committed:
Remove uneeded travis.yml setup

- increase install performance
- fix bug in copy source

noahtallen committed:
Minor fixes

noahtallen committed:
revert permformance test

noahtallen committed:
remove old stuf fagain

epiqueras reviewed (approved):
Nice work.

Let's stay alert for any bugs before the next release.

noahtallen committed:
Fixes for config tests

noahtallen committed:
Fix docker compose tests

noahtallen committed:
Update changelog

noahtallen committed:
Only map e2e plugins to test environment

noahtallen committed:
Update documentation

noahtallen committed:
typo

epiqueras commented on packages/env/README.md:
```diff
@@ -284,17 +284,16 @@ Options:
+| `"core"`     | `string\|null` | `null`                                 | The WordPress installation to use. If `null` is specified, `wp-env` will use the latest production release of WordPress.  |
+| `"plugins"`  | `string[]`     | `[]`                                   | A list of plugins to install and activate in the environment.                                                             |
+| `"themes"`   | `string[]`     | `[]`                                   | A list of themes to install in the environment. The first theme in the list will be activated.                            |
+| `"port"`     | `integer`      | `8888` (`8889` for the tests instance) | The primary port number to use for the insallation. You'll access the instance through the port: 'http://localhost:8888'. |
```
```suggestion
| `"port"`     | `integer`      | `8888` (`8889` for the tests instance) | The primary port number to use for the installation. You'll access the instance through the port: 'http://localhost:8888'. |
```

epiqueras commented on packages/env/README.md:
```diff
@@ -309,9 +308,59 @@ Several types of strings can be passed into the `core`, `plugins`, `themes`, and
 
 Remote sources will be downloaded into a temporary directory located in `~/.wp-env`.
 
+Additionally, the key `env` is available to override any of the above options on an individual-environment basis. For example, take following `.wp-env.json` file:
```
```suggestion
Additionally, the key `env` is available to override any of the above options on an individual-environment basis. For example, take the following `.wp-env.json` file:
```

epiqueras commented on packages/env/README.md:
```diff
@@ -309,9 +308,59 @@ Several types of strings can be passed into the `core`, `plugins`, `themes`, and
 ## .wp-env.override.json
 
-Any fields here will take precedence over .wp-env.json. This file is useful, when ignored from version control, to persist local development overrides.
+Any fields here will take precedence over .wp-env.json. This file is useful, when ignored from version control, to persist local development overrides. Note that options like `plugins` and `themes` are not merged. As a result, if you set `plugins` in your override file, this will override all of the plugins listed in the base-level config. The only keys whch are merged are `config` and `mappings` ...
```
```suggestion
Any fields here will take precedence over .wp-env.json. This file is useful when ignored from version control, to persist local development overrides. Note that options like `plugins` and `themes` are not merged. As a result, if you set `plugins` in your override file, this will override all of the plugins listed in the base-level config. The only keys which are merged are `config` and `mappings`. This means that you can set your own wp-config values without losing any of the default values.
```

epiqueras commented:
The docs are getting a bit redundant. We might want to lean them out in a follow-up.

noahtallen committed:
Fix more typos

Thanks macbook

Co-authored-by: Enrique Piqueras <epiqueras@users.noreply.github.com>

noahtallen commented:
I definitely agree. I bet it's tough for people to find what they need.

noahtallen commented:
The remaining task here is to introduce unit tests to cover the new functionality. Fixing the existing unit tests uncovered several subtle bugs which are now resolved. Thanks unit tests 💯

noahtallen committed:
Remove overrides from github workflows

noahtallen committed:
Add tests for back compatibility and file reading

noahtallen committed:
Mappings, config, overrides, merging tests

noahtallen committed:
Test source override behavior
